frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+image datasets. To do so, one may freeze the pretrained model, utilizing
+it as a feature extractor to train a much smaller model that is built on
+top of the feature extraction. One can also fine tune the entire model
+to fit the new task. Transfer learning has a series of challenges, in
+that the modified model may not be able to conduct its original tasks
+after transfer learning. Papers such as [["Learning without
+Forgetting"]{.underline}](https://browse.arxiv.org/pdf/1606.09282.pdf)
+paper aims to address these challenges and have been implemented in
+modern machine learning platforms.
+
+##### Federated Learning
 
-## Embedded AI Constraints
+Consider the problem of labeling items that are present in a photo from
+personal devices. One may consider moving the image data from the
+devices to a central server, where a single model will train Using these
+image data provided by the devices. However, this presents many
+potential challenges. First, with many devices one needs a massive
+network infrastructure to move and store data from these devices to a
+central location. With the number of devices that are present today this
+is often not feasible, and very costly. Furthermore, there are privacy
+challenges associated with moving personal data, such as Photos central
+servers.
 
-Explanation: Describe the constraints of embedded systems, referring to the previous chapters, and remind readers about the challenges and why we need to consider creating lean and efficient solutions.
+[[Federated learning]{.underline}](https://arxiv.org/abs/1602.05629) is
+a form of distributed computing that resolves these issues by
+distributing the models into personal devices for them to be trained on
+device. At the beginning, a base global model is trained on a central
+server to be distributed to all devices. Using this base model, the
+devices individually compute the gradients and send them back to the
+central hub. Intuitively this is the transfer of model parameters
+instead of the data itself. This innovative approach allows the model to
+be trained with many different datasets (which, in our example, would be
+the set of images that are on personal devices), without the need to
+transfer a large amount of potentially sensitive data. However,
+federated learning also comes with a series of challenges.
 
-### Hardware
+Training in machine learning usually assumes the data to be identically
+and independently distributed. However, the nature of federated learning
+makes it so that this is not necessarily the case. In our example,
+devices may have similar photos from being at the same location, and the
+types of the cameras that each of the devices have may be different.
+There are many different techniques to compensate for this, such as
+adding a proximal term to achieve a balance between the local and global
+model, and adding a frozen [[global hypersphere
+classifier]{.underline}](https://arxiv.org/abs/2207.09413).
```

Comment:
Labeled Data is Not Always Naturally Available:
- In many real-world scenarios, data collected on devices might not come with appropriate labels.
- Additionally, users, being a source of data, can be unreliable, meaning that even if data is labeled, there's no guarantee of its accuracy or relevance.

Non-IID Data (covered already, but can be made clearer):
- Each user's data is unique, leading to significant variance in the kind of data generated by different users.
- The unbalanced nature of data, where some users produce more data than others can affect the global model's performance.

Massively Distributed:
- The number of mobile device owners can be significantly larger than the average number of training samples on each device. This leads to a significant communication overhead.

Limited Communication:
- Mobile networks, which are typically used for such communication, can be unstable. This instability can lead to delayed or failed transmission of model updates, affecting the overall training process.

Heterogeneous Device/Silo Resources:
- Devices participating in FL can have varying computational powers and memory capacities, making it challenging to design algorithms that are efficient across all devices.

Privacy, Security, and Trust:
- Malicious clients or servers can attempt to reverse-engineer model updates to infer information about local data.
- With malicious clients or servers in the system, there's a risk of model poisoning or other types of adversarial attacks.

Reply:
That's excellent feedback. Thanks for providing that.

Reply:
Got it. Thanks!